README
Document library project

To install:
- clone this one
- install joomla from this one like normaly
- create tables for com_document library
- please, don't commit the deletion of installation folder



installation_/CHANGELOG
0. Metadata
-----------
File version:	$Id: CHANGELOG 21693 2011-06-27 16:44:53Z dextercowley $
Copyright:		Copyright (C) 2005 - 2011 Open Source Matters, Inc. All rights reserved.
License:		GNU General Public License version 2 or later; see LICENSE.txt

1. Copyright and disclaimer
---------------------------
This application is opensource software released under the GPL.  Please
see source code and the LICENSE file


2. Changelog
------------
This is a non-exhaustive (but still near complete) changelog for
Joomla! 1.6, including alpha and beta versions.
Our thanks to all those people who've contributed bug reports and
code fixes.

Legend:

* -> Security Fix
# -> Issue Fix
$ -> Language fix or change
+ -> Addition
^ -> Change
- -> Removed
! -> Note

-------------------- 1.6.4 Stable Release [27-Jun-2011] ------------------

27-Jun-2011 Mark Dexter
 # [#25696] Updating from J1.6.2 to J1.6.3 removes all files!

21-Jun-2011 Mark Dexter
 # [#26171] Automatic DB update (Christophe Demko)

-------------------- 1.6.3 Stable Release [18-Apr-2011] ------------------

18-Apr-2011 Mark Dexter
 # [#25661] *Redirect not now fully editable
 # [#25657] Mootools no longer loaded in backend
 # [#25688] Updates sites never re-enabled after being disabled

16-Apr-2011 Jean-Marie Simonet
 $ Updating es-ES installation file
 # [#25650] Can't insert image for menu item. Thanks Dennis and Rouven
 
15-Apr-2011 Jean-Marie Simonet
 # Removing install check

-------------------- 1.6.2 Stable Release [14-Apr-2011] ------------------

14-Apr-2011
 # [#25617] performance of assets table query with > 40 articles on website in registered mode

13-Apr-2011 Jean-Marie Simonet
 # Correcting error in the plugin language filter xml
 
12-Apr-2011 Mark Dexter
 # [#25593] Module Articles Category takes server timezone instead of Joomla one
 # [#25370] new layouts - No pagination on articles category page if first visit

11-Apr-2011 Mark Dexter
 # [#25602] Sort in Article Category List broken
 # [#23364] Installation of tar.gz packages fails in xampp under Windows
 # [#25614] *en-GB.files_joomla.sys.ini not loading
 # [#25541] Apply (Save and present in Editor again) does not checkout row
 # [#23878] Remove remaining ereg* functions - Obscolete in PHP 5.3
 # [#25581] RTL in Atomic
 # [#25603] *Email Cloak Plugin outputs preformatted code in email "to" field
 # [#21809] access levels - pagination::orderDownIcon appearing before last item
 # [#24835] Administrator (with a capital A) does weird stuff

10-Apr-2011 Jean-Marie Simonet
 $ Adding lib_joomla.ini into site language folder and localise.php in admin language folder to prevent confusion for TTs.
 $ Changing en-GB.xmls version for 1.6.2
 
09-Apr-2011 Jean-Marie Simonet
 #$ Updating pl-PL installation language

08-Apr-2011 Jean-Marie Simonet
 # [#25585] *Unused Jtoolbarhelper title in Language Manager Content Languages
 # [#24321] Hiding "Add New Shortcuts" removed Groups and Levels in admin menu. Thanks Christophe.
 # [#25257] media manager folder names hidden when long. Thanks Bill and Andrea.

07-Apr-2011
 # [#24016] Native support for IIS Rewrite
 # [#25247] Fix a number of incorrect comment in components
 # [#25563] Category Blog - More Article links don't have limitation
 # [#25497] *Notices with module Articles category
 # [#25576] Gmail authentication fails (Verify Peer variable mistyped)
 # [#25550] com_contact vCard link not using JRoute::_() which causes download issues with System - Language Filter plugin enabled and SEF

07-Apr-2011 Jean-Marie Simonet
 # [#25569] the package strings are not translated when installing or managing. Thanks Christophe

06-Apr-2011 Jean-Marie Simonet
 # [#25551] #24285 change breaks module cache functionality, needs to be reverted. Thanks Klas.
 # [#25572] *ROOT parent category wrongly displays in featured and blog layout
 # [#25571] admin templates does not add a border on fielset when using tabs. Thanks Christophe.

05-Apr-2011 Mark Dexter
 # [#25540] JModelForm enables 'com_checkin' for 'core.manage' - should be 'core.admin'
 # [#24974] The word Array appears below the document when Cache is turned on
 # [#25545] Wrong header output in com_contact vCard view makes IE handle vCard downloads wrong...

05-Apr-2011 Jean-Marie Simonet
 #$ Adding Icelandic to installation languages
 #$ Upgrading ru_RU installation language
 
04-Apr-2011 Mark Dexter
 # [#25519] Minor Bugs in mod_whosonline

04-Apr-2011 Jean-Marie Simonet
 #$ Upgrading fi-FI installation language

02-Apr-2011 Mark Dexter
 # [#25340] The robots.txt file offers no protection for a site installed in a folder.
 # [#24285] Headlinks in header sum up when using progressive/conservative cache
 # [#25403] "Font Size" not working with Internet Explorer 7
 # [#25508] Update CodeMirror to Version 1.0
 # [#22870] Improved fix for XSS issue found in Joomla 1.5.20
 # [#25522] Extra quote in com_weblinks
 # [#25074] Form onchange event pointing to non-existant event
 # [#22926] browser detection script reports wrong informations
 # [#25397] Typos in sample data
 
02-Apr-2011 Jean-Marie Simonet
 $ Added missing strings

01-Apr-2011 Mark Dexter
 # Fix incorrect comments in articles model

31-Mar-2011 Mark Dexter
 # [#24917] Module latest user parameter link to contact does not work
 # [#25385] Optimize for loops
 # [#25425] Extension "upgrade" method re-installing component
 # [#25496] Database error not handled in mod_users_latest and mod_articles_archive

31-Mar-2011 Jean-Marie Simonet
 # [#25495] *Wrong lang load order when editing plugin
 # Normalizing plugins xml (languages missing). Thanks Bill
 # [#25486] Undefined notice on 404 page

30-Mar-2011 Jean-Marie Simonet
 #$ Adding Bosnian to installation languages
 
29-Mar-2011 Mark Dexter
 # [#25288] Media form field is not allowed access to media manager in the front end
 # [#25129] Media manager popup - Change of directory don't reflect in list of images and folders on windows system (XAMPP)

plugins/user/profile/profile.php
<?php
/**
 * @version		$Id: profile.php 21097 2011-04-07 15:38:03Z dextercowley $
 * @copyright	Copyright (C) 2005 - 2011 Open Source Matters, Inc. All rights reserved.
 * @license		GNU General Public License version 2 or later; see LICENSE.txt
 */

defined('JPATH_BASE') or die;
jimport('joomla.utilities.date');

/**
 * An example custom profile plugin.
 *
 * @package		Joomla.Plugin
 * @subpackage	User.profile
 * @version		1.6
 */
class plgUserProfile extends JPlugin
{
	/**
	 * @param	string	$context	The context for the data
	 * @param	int		$data		The user id
	 * @param	object
	 *
	 * @return	boolean
	 * @since	1.6
	 */
	function onContentPrepareData($context, $data)
	{
		// Check we are manipulating a valid form.
		if (!in_array($context, array('com_users.profile','com_users.user', 'com_users.registration', 'com_admin.profile'))) {
			return true;
		}

		if (is_object($data))
		{
			$userId = isset($data->id) ? $data->id : 0;

			// Load the profile data from the database.
			$db = JFactory::getDbo();
			$db->setQuery(
				'SELECT profile_key, profile_value FROM #__user_profiles' .
				' WHERE user_id = '.(int) $userId." AND profile_key LIKE 'profile.%'" .
				' ORDER BY ordering'
			);
			$results = $db->loadRowList();

			// Check for a database error.
			if ($db->getErrorNum())
			{
				$this->_subject->setError($db->getErrorMsg());
				return false;
			}

			// Merge the profile data.
			$data->profile = array();

			foreach ($results as $v)
			{
				$k = str_replace('profile.', '', $v[0]);
				$data->profile[$k] = $v[1];
			}
			if (!JHtml::isRegistered('users.url')) {
				JHtml::register('users.url', array(__CLASS__, 'url'));
			}
			if (!JHtml::isRegistered('users.calendar')) {
				JHtml::register('users.calendar', array(__CLASS__, 'calendar'));
			}
			if (!JHtml::isRegistered('users.tos')) {
				JHtml::register('users.tos', array(__CLASS__, 'tos'));
			}
			if (!JHtml::isRegistered('users.degree')) {
				JHtml::register('users.degree', array(__CLASS__, 'degree'));
			}
			if (!JHtml::isRegistered('users.subject')) {
				JHtml::register('users.subject', array(__CLASS__, 'subject'));
			}
		}

		return true;
	}

	public static function url($value)
	{
		if (empty($value))
		{
			return JHtml::_('users.value', $value);
		}
		else
		{
			$value = htmlspecialchars($value);
			if(substr ($value, 0, 4) == "http") {
				return '<a href="'.$value.'">'.$value.'</a>';
			}
			else {
				return '<a href="http://'.$value.'">'.$value.'</a>';
			}
		}
	}

	public static function calendar($value)
	{
		if (empty($value)) {
			return JHtml::_('users.value', $value);
		} else {
			return JHtml::_('date', $value);
		}
	}

	public static function tos($value)
	{
		if ($value) {
			return JText::_('JYES');
		}
		else {
			return JText::_('JNO');
		}
	}
	
	public static function degree($value) {
		$values= array();
		$values[1] = "PLG_USER_PROFILE_FIELD_DEGREE_BACHELOR_OPTION";
		$values[2] = "PLG_USER_PROFILE_FIELD_DEGREE_MASTER_OPTION";
		$values[3] = "PLG_USER_PROFILE_FIELD_DEGREE_DOCTOR_OPTION";
		if (empty($value)) {
			return JHtml::_('users.value', $value);
		}
		if (array_key_exists($value, $values)) {
			return JText::_($values[$value]);
		}
		return JText::_($value);
	} 
	
	public static function subject($value) {
		if (empty($value)) {
			return JHtml::_('users.value', $value);
		}
		static $languageLoaded = false;
		if (!$languageLoaded) {
			$language =& JFactory::getLanguage();
			$extension = 'com_documentlibrary';
			$base_dir = JPATH_SITE . '/components/com_documentlibrary';
			$language_tag = $language->getTag(); // loads the current language-tag
			$language->load($extension, $base_dir, $language_tag, true);
			$languageLoaded = true;
		}
		$db = JFactory::getDbo();
		$query = 'SELECT name FROM #__document_subjects WHERE subject_id = ' . $value;
		$db->setQuery($query);
		$name = $db->loadResult();
		return JText::_($name);
	}

	/**
	 * @param	JForm	$form	The form to be altered.
	 * @param	array	$data	The associated data for the form.
	 *
	 * @return	boolean
	 * @since	1.6
	 */
	function onContentPrepareForm($form, $data)
	{
		// Load user_profile plugin language
		$lang = JFactory::getLanguage();
		$lang->load('plg_user_profile', JPATH_ADMINISTRATOR);

		if (!($form instanceof JForm))
		{
			$this->_subject->setError('JERROR_NOT_A_FORM');
			return false;
		}

		// Check we are manipulating a valid form.
		if (!in_array($form->getName(), array('com_admin.profile','com_users.user', 'com_users.registration','com_users.profile'))) {
			return true;
		}

		// Add the registration fields to the form.
		JForm::addFormPath(dirname(__FILE__).'/profiles');
		$form->loadFile('profile', false);

		// Toggle whether the address1 field is required.
		if ($this->params->get('register-require_address1', 1) > 0) {
			$form->setFieldAttribute('address1', 'required', $this->params->get('register-require_address1') == 2, 'profile');
		}
		else {
			$form->removeField('address1', 'profile');
		}

		// Toggle whether the address2 field is required.
		if ($this->params->get('register-require_address2', 1) > 0) {
			$form->setFieldAttribute('address2', 'required', $this->params->get('register-require_address2') == 2, 'profile');
		}
		else {
			$form->removeField('address2', 'profile');
		}

		// Toggle whether the city field is required.
		if ($this->params->get('register-require_city', 1) > 0) {
			$form->setFieldAttribute('city', 'required', $this->params->get('register-require_city') == 2, 'profile');
		}
		else {
			$form->removeField('city', 'profile');
		}

		// Toggle whether the region field is required.
		if ($this->params->get('register-require_region', 1) > 0) {
			$form->setFieldAttribute('region', 'required', $this->params->get('register-require_region') == 2, 'profile');
		}
		else {
			$form->removeField('region', 'profile');
		}

		// Toggle whether the country field is required.
		if ($this->params->get('register-require_country', 1) > 0) {
			$form->setFieldAttribute('country', 'required', $this->params->get('register-require_country') == 2, 'profile');
		}
		else {
			$form->removeField('country', 'profile');
		}

		// Toggle whether the postal code field is required.
		if ($this->params->get('register-require_postal_code', 1) > 0) {
			$form->setFieldAttribute('postal_code', 'required', $this->params->get('register-require_postal_code') == 2, 'profile');
		}
		else {
			$form->removeField('postal_code', 'profile');
		}

		// Toggle whether the phone field is required.
		if ($this->params->get('register-require_phone', 1) > 0) {
			$form->setFieldAttribute('phone', 'required', $this->params->get('register-require_phone') == 2, 'profile');
		}
		else {
			$form->removeField('phone', 'profile');
		}

		// Toggle whether the website field is required.
		if ($this->params->get('register-require_website', 1) > 0) {
			$form->setFieldAttribute('website', 'required', $this->params->get('register-require_website') == 2, 'profile');
		}
		else {
			$form->removeField('website', 'profile');
		}

		// Toggle whether the favoritebook field is required.
		if ($this->params->get('register-require_favoritebook', 1) > 0) {
			$form->setFieldAttribute('favoritebook', 'required', $this->params->get('register-require_favoritebook') == 2, 'profile');
		}
		else {
			$form->removeField('favoritebook', 'profile');
		}

		// Toggle whether the aboutme field is required.
		if ($this->params->get('register-require_aboutme', 1) > 0) {
			$form->setFieldAttribute('aboutme', 'required', $this->params->get('register-require_aboutme') == 2, 'profile');
		}
		else {
			$form->removeField('aboutme', 'profile');
		}

		// Toggle whether the tos field is required.
		if ($this->params->get('register-require_tos', 1) > 0) {
			$form->setFieldAttribute('tos', 'required', $this->params->get('register-require_tos') == 2, 'profile');
		}
		else {
			$form->removeField('tos', 'profile');
		}

		// Toggle whether the dob field is required.
		if ($this->params->get('register-require_dob', 1) > 0) {
			$form->setFieldAttribute('dob', 'required', $this->params->get('register-require_dob') == 2, 'profile');
		}
		else {
			$form->removeField('dob', 'profile');
		}

		// $oo = $form->getField('degree', 'profile');
		// var_dump($oo);
		// $subjectField = $this->createSubjectField();
		// var_dump($subjectField);
		// $form->setField($subjectField, 'profile', true);

		return true;
	}

	function onUserAfterSave($data, $isNew, $result, $error)
	{
		$userId	= JArrayHelper::getValue($data, 'id', 0, 'int');

		if ($userId && $result && isset($data['profile']) && (count($data['profile'])))
		{
			try
			{
				//Sanitize the date
				if (!empty($data['profile']['dob'])) {
					$date = new JDate($data['profile']['dob']);
					$data['profile']['dob'] = $date->toFormat('%Y-%m-%d');
				}

				$db = JFactory::getDbo();
				$db->setQuery(
					'DELETE FROM #__user_profiles WHERE user_id = '.$userId .
					" AND profile_key LIKE 'profile.%'"
				);

				if (!$db->query()) {
					throw new Exception($db->getErrorMsg());
				}

				$tuples = array();
				$order	= 1;

				foreach ($data['profile'] as $k => $v)
				{
					$tuples[] = '('.$userId.', '.$db->quote('profile.'.$k).', '.$db->quote($v).', '.$order++.')';
				}

				$db->setQuery('INSERT INTO #__user_profiles VALUES '.implode(', ', $tuples));

				if (!$db->query()) {
					throw new Exception($db->getErrorMsg());
				}

			}
			catch (JException $e)
			{
				$this->_subject->setError($e->getMessage());
				return false;
			}
		}

		return true;
	}

	/**
	 * Remove all user profile information for the given user ID
	 *
	 * Method is called after user data is deleted from the database
	 *
	 * @param	array		$user		Holds the user data
	 * @param	boolean		$success	True if user was succesfully stored in the database
	 * @param	string		$msg		Message
	 */
	function onUserAfterDelete($user, $success, $msg)
	{
		if (!$success) {
			return false;
		}

		$userId	= JArrayHelper::getValue($user, 'id', 0, 'int');

		if ($userId)
		{
			try
			{
				$db = JFactory::getDbo();
				$db->setQuery(
					'DELETE FROM #__user_profiles WHERE user_id = '.$userId .
					" AND profile_key LIKE 'profile.%'"
				);

				if (!$db->query()) {
					throw new Exception($db->getErrorMsg());
				}
			}
			catch (JException $e)
			{
				$this->_subject->setError($e->getMessage());
				return false;
			}
		}

		return true;
	}
	
	private function getDataAgain($userId, &$data)
	{
		if ($userid <= 0) {
			return null;
		}

			$data	= new JUser($userId);

			// Set the base user data.
			$data->email1 = $data->get('email');
			$data->email2 = $data->get('email');

			// Override the base user data with any data in the session.
			$temp = (array)JFactory::getApplication()->getUserState('com_users.edit.profile.data', array());
			foreach ($temp as $k => $v) {
				$data->$k = $v;
			}

			// Unset the passwords.
			unset($data->password1);
			unset($data->password2);

			$registry = new JRegistry($data->params);
			$data->params = $registry->toArray();

			// Get the dispatcher and load the users plugins.
			// $dispatcher	= JDispatcher::getInstance();
			// JPluginHelper::importPlugin('user');
// 
			// // Trigger the data preparation event.
			// $results = $dispatcher->trigger('onContentPrepareData', array('com_users.profile', $this->data));
// 
			// // Check for errors encountered while preparing the data.
			// if (count($results) && in_array(false, $results, true)) {
				// $this->setError($dispatcher->getError());
				// $this->data = false;
			// }
		// }

		return $data;
	}
}
